# Pipeline for CH12F3

## samples in this part
| sample title | pipeline |
| :---: | :---: |
| MM Bcell AK rep1 | CH12F3 |
| MM Bcell AK rep2 | CH12F3 |
| MM Bcell UK rep1 | CH12F3 |
| MM Bcell UK rep2 | CH12F3 |


Deduplicates:
```bash
mm_samples=(MM_AK_rep1 MM_AK_rep2 MM_UK_rep1 MM_UK_rep2)
for sm in ${mm_samples[@]}
do
	picard MarkDuplicates --INPUT ${sm}_sorted.bam --output ${sm}_sorted_dedup.bam --METRICS_FILE ${sm}_sorted_dedup.metrics \
	--REMOVE_DUPLICATES true --CREATE_INDEX true --VALIDATION_STRINGENCY SILENT
done

```

We only intrested in AID-hotspot region, therefore the information of mapped reads within **chr12:113418578-113427207** was acquired by run
```bash
for sm in ${mm_samples[@]}
do
	python getInfoinAID.py --ibam ${sm}_sorted_dedup.bam --region chr12:113418578-113427207 --genome mm10.fa --out ${sm}.inAID.txt
done

``` 
